File: Objects/PipeSet.py
import pygame
import random
import math
from pathlib import PurePath
from Objects.ColorPalette import ColorPalette
from Objects.Sprite import Sprite


class PipeSet:
    """
    Represents an obstacle in the game
    """
    def __init__(self):
        """
        Initializes a new instance of the pipe class
        """
        self.canvas_height, self.canvas_width = pygame.display.get_surface().get_size()
        self.color_palette = ColorPalette()

        self.x_coordinate = self.canvas_height
        self.passable_space_height = 30
        self.pipe_width = 60
        self.offset = self.pipe_width // 2 # Divide the pipe's width then floor (//) the result

        # Sprites
        self.__pipe_head = Sprite(str(PurePath("res/Images/pipe_head.png")))
        self.__pipe_body = Sprite(str(PurePath("res/Images/pipe_body.png")))

        self.top_pipe_height = 0
        self.bottom_pipe_height = 0
        self.scroll_speed = 2  # How fast the pipes will move towards the bird

        self.calculate_dimensions()

    def to_canvas(self, canvas):
        """
        Draws the bird onto the specified canvas or surface \n
        :param canvas: The surface wherein the bird is to be drawn on
        """
        # top pipe
        # pygame.draw.line(canvas, self.color_palette.white, (self.x_coordinate, 0),
        #                 (self.x_coordinate, self.top_pipe_height), self.pipe_width)
        # bottom pipe
        # pygame.draw.line(canvas, self.color_palette.white, (self.x_coordinate, self.bottom_pipe_height),
        #              (self.x_coordinate, self.canvas_height*2), self.pipe_width)

        # The pipes have the same dimensions for the head sprite
        pipe_head_dimensions = (self.pipe_width, self.offset)

        # Parameters for the top pipe
        top_pipe_body_location = ((self.x_coordinate - self.offset), -self.offset)
        top_pipe_head_location = ((self.x_coordinate - self.offset), (self.top_pipe_height - self.offset))
        top_pipe_body_dimensions = (self.pipe_width, self.top_pipe_height)

        # Parameters for the bottom pipe
        bottom_pipe_body_location = ((self.x_coordinate - self.offset), self.bottom_pipe_height)
        bottom_pipe_head_location = ((self.x_coordinate - self.offset), (self.bottom_pipe_height - self.offset))
        bottom_pipe_body_dimensions = (self.pipe_width, self.bottom_pipe_height*2)

        # Draw the top pipe
        self.__pipe_body.to_canvas(canvas=canvas, location=top_pipe_body_location,
                                   dimensions=top_pipe_body_dimensions)
        self.__pipe_head.to_canvas(canvas=canvas, location=top_pipe_head_location,
                                   dimensions=pipe_head_dimensions)

        # Draw the bottom pipe
        self.__pipe_body.to_canvas(canvas=canvas, location=bottom_pipe_body_location,
                                   dimensions=bottom_pipe_body_dimensions)
        self.__pipe_head.to_canvas(canvas=canvas, location=bottom_pipe_head_location,
                                   dimensions=pipe_head_dimensions)

    def calculate_dimensions(self):
        """
        Calculates the height and gap between the two pipes in the pipe set
        """
        max_height = (self.canvas_height // 2) - self.passable_space_height
        self.top_pipe_height = random.choice(range(self.passable_space_height, max_height))
        self.bottom_pipe_height = random.choice(range(max_height, self.canvas_height))

    def scroll(self):
        """
        Gradually move the pipe towards the bird and off the screen
        """
        self.x_coordinate -= self.scroll_speed

    def collide(self, bird):
        """
        Determines whether the bird had collided with either of the two pipes in the pipe set \n
        :param bird: The bird object that the player controls \n
        :return:True if the bird had collided with any of the pipes in the set
        """
        bird_inside_pipe_area = (bird.x_coordinate >= self.x_coordinate) \
                                and (bird.x_coordinate <= self.x_coordinate + self.pipe_width)
        bird_in_contact = (bird.y_coordinate <= self.top_pipe_height) \
                          or (bird.y_coordinate >= self.bottom_pipe_height)

        if bird_in_contact and bird_inside_pipe_area:
            return True
        else:
            return False

    def is_cleared(self, bird):
        """
        Determines whether the bird has successfully passed through the pipe \n
        :return: True if the bird did not hit any of the pipes in the pipe set
        """
        mid_point_x = (self.x_coordinate + bird.x_coordinate) // 2
        mid_point_y = (self.top_pipe_height + self.bottom_pipe_height) // 2

        # if the bird passes through the are in-between the two pipes, it has cleared the obstacle
        if (bird.x_coordinate == mid_point_x and bird.y_coordinate >= mid_point_y)\
                or (bird.x_coordinate == mid_point_x and bird.y_coordinate <= mid_point_y):
            return True
        else:
            return False

File: docs/GameObjects/PipeSet.js
/**
 * Represents an in-game obstacle
 * @class PipeSet
 */
class PipeSet {

    /**
     * Creates an instance of PipeSet.
     * @memberof PipeSet
     */
    constructor() {
        this.x_coordinate =  innerWidth;
        this.pipeWidth = 80;
        this.passableSpaceHeight = this.pipeWidth;
        this.offSet = this.pipeWidth / 2;

        this.pipeHead = new Sprite("res/Images/pipe_head.png");
        this.pipeBody = new Sprite("res/Images/pipe_body.png");

        this.topPipeHeight = 0;
        this.bottomPipeHeight = 0;
        this.cleared = false;

        this.calculateDimensions();
    }

    /**
     * Calculates the height and gap between the two pipes in the pipe set
     * @memberof PipeSet
     */
    calculateDimensions() {
        let maxHeight = Math.abs((innerHeight / 2) - this.passableSpaceHeight * 4);
        // max defines how small the bottom pipe would be since this one grows from top to bottom
        this.bottomPipeHeight = this.getRandom(maxHeight, innerHeight - maxHeight);
        this.topPipeHeight = this.getRandom(this.passableSpaceHeight, maxHeight);        
    }


    /**
     * Gradually move the pipe set towards the bird and off the screen
     * @param {number} [scrollSpeed=2] How fast the pipe set will go towards the bird
     * @memberof PipeSet
     */
    scroll(scrollSpeed=2) {
        this.x_coordinate -= scrollSpeed;
    }

    /**
     * Draws the object's sprite onto the canvas
     * @memberof PipeSet
     */
    toCanvas() {

        // Pipe head parameters
        let pipeHeadWidth = this.pipeWidth;
        let pipeHeadHeight = this.offSet;

        // Top pipe body parameters
        let topPipeBody_xCoordinate = (this.x_coordinate - this.offSet);
        let topPipeBody_yCoordinate = -this.offSet;
        // Top pipe head paramters
        let topPipeHead_xCoordinate = (this.x_coordinate - this.offSet);
        let topPipeHead_yCoordinate = (this.topPipeHeight - this.offSet);
        // Top pipe dimensions
        let d_topPipeWidth = this.pipeWidth;
        let d_topPipeHeight = this.topPipeHeight;

        // Bottom pipe body parameters
        let bottomPipeBody_xCoordinate = (this.x_coordinate - this.offSet);
        let bottomPipeBody_yCoordinate = this.bottomPipeHeight + this.offSet;
        // Bottom pipe head paramters
        let bottomPipeHead_xCoordinate = (this.x_coordinate - this.offSet);
        let bottomPipeHead_yCoordinate = (this.bottomPipeHeight + this.offSet);
        // Bottom pipe dimensions
        let d_bottomPipeWidth = this.pipeWidth;
        let d_bottomPipeHeight = this.bottomPipeHeight * 4;
        
        // Draw the top pipe
        this.pipeBody.toCanvas(topPipeBody_xCoordinate, topPipeBody_yCoordinate, 
            d_topPipeHeight, d_topPipeWidth);
        this.pipeHead.toCanvas(topPipeHead_xCoordinate, topPipeHead_yCoordinate, 
            pipeHeadHeight, pipeHeadWidth);
        
        // Draw the bottom pipe
        this.pipeBody.toCanvas(bottomPipeBody_xCoordinate, bottomPipeBody_yCoordinate, 
            d_bottomPipeHeight, d_bottomPipeWidth);
        this.pipeHead.toCanvas(bottomPipeHead_xCoordinate, bottomPipeHead_yCoordinate, 
            pipeHeadHeight, pipeHeadWidth);
    }

    /**
     * Determine whether the bird has hit any of the pipes in the pipe set.
     * @param {Bird} bird the instance of the bird class in the game world.
     * @returns A boolean value based on whether the bird had hit (true) any of the pipes
     *          or not (false).
     * @memberof PipeSet
     */
    collide(bird) {
        let bird_in_contact = ((this.x_coordinate - this.pipeWidth) - 
            bird.x_coordinate) <= 0; 
        let midPoint_y = ((this.topPipeHeight + this.bottomPipeHeight) / 2) - 
            this.pipeWidth;
        let isOverPipes = 
            (
                ((bird.y_coordinate <= this.topPipeHeight) 
                && (bird.y_coordinate < midPoint_y)) ||
                ((bird.y_coordinate >= this.bottomPipeHeight)
                && (bird.y_coordinate > midPoint_y))
            );
        
        if (bird_in_contact && isOverPipes) {
                return true;
            }
            else {
                return false;
            }
    }

    /**
     * Determine whether the bird has successfully passed through the open space
     * in-between the two pipes in the pipe set.
     * 
     * @param {Bird} bird the instance of the bird class in the game world.
     * @returns A boolean value based on whether the bird had successfully passed through
     *          the two pipes (true) or not (false).
     * @memberof PipeSet
     */
    isCleared(bird) {        
        let isPastPipes = bird.x_coordinate > this.x_coordinate;
        let isInBetweenPipes = 
            (
                (bird.y_coordinate > this.topPipeHeight) ||
                (bird.y_coordinate < this.bottomPipeHeight)
            );

        if (!this.cleared && (isPastPipes && isInBetweenPipes)) {
            this.cleared = true;
            return this.cleared;
        }
        else {
            return false;
        }
    }


    /**
     * Reconstruct the pipes in the pipe set.
     * @memberof PipeSet
     */
    reconstruct() {
        this.x_coordinate = innerWidth;
        this.calculateDimensions();
        this.cleared = false;
    }

    /**
     * Returns a random number between the given intervals min and max.
     *
     * @param {Number} min the lower bound of the selection (exclusive).
     * @param {Number} max the upper bound of the selection (inclusive).
     * @returns a random Number in-between the given boundaries.
     * @memberof PipeSet
     */
    getRandom(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
}